File: scripts/governance/report-metrics.js
#!/usr/bin/env node
/**
 * Governance Metrics Report — renders .memory-layer/metrics/latest.json as Markdown.
 */

const fs = require('fs');

const METRICS_FILE = '.memory-layer/metrics/latest.json';

if (!fs.existsSync(METRICS_FILE)) {
  console.log('No metrics yet. Run: node scripts/governance/collect-metrics.js');
  process.exit(0);
}

const m = JSON.parse(fs.readFileSync(METRICS_FILE, 'utf8'));

const THRESHOLDS = {
  any_count: { limit: 0, label: 'TypeScript `any` usage' },
  eslint_disable: { limit: 0, label: 'eslint-disable suppressions' },
  silent_catches: { limit: 0, label: 'Silent catch blocks' },
  python_type_holes: { limit: 0, label: 'Python type holes' },
  python_noqa: { limit: 0, label: 'Python noqa suppressions' },
};

const rows = Object.entries(THRESHOLDS).map(([key, { limit, label }]) => {
  const val = m[key] ?? 0;
  const status = val <= limit ? '✓' : '✗';
  return `| ${status} | ${label} | ${val} | ${limit} |`;
});

const report = `
# Governance Metrics Report

Generated: ${m.collected_at}

## Quality Gates

| Status | Metric | Current | Limit |
|--------|--------|---------|-------|
${rows.join('\n')}

## Codebase Size

| Metric | Value |
|--------|-------|
| Source LOC (TS) | ${m.source_loc} |
| Test LOC | ${m.test_loc} |
| Mock Tax Ratio | ${m.source_loc > 0 ? (m.test_loc / m.source_loc).toFixed(2) : 'N/A'}× |
`.trim();

console.log(report);

const outFile = '.memory-layer/metrics/report.md';
fs.writeFileSync(outFile, report + '\n');
console.log(`\nReport written to ${outFile}`);
process.exit(0);

File: scripts/governance/report_metrics.py
#!/usr/bin/env python3
"""Governance Metrics Report — render latest metrics as Markdown"""

import json
import sys
from pathlib import Path

METRICS_FILE = Path(".memory-layer/metrics/latest.json")

if not METRICS_FILE.exists():
    print("No metrics yet. Run: python3 scripts/governance/collect_metrics.py")
    sys.exit(0)

m = json.loads(METRICS_FILE.read_text())

THRESHOLDS = {
    "ts_any_count": (0, "TypeScript `any` usage"),
    "ts_eslint_disable": (0, "eslint-disable suppressions"),
    "python_type_holes": (0, "Python type holes"),
    "python_noqa": (0, "Python noqa suppressions"),
}

rows = []
for key, (limit, label) in THRESHOLDS.items():
    val = m.get(key, 0)
    status = "✓" if val <= limit else "✗"
    rows.append(f"| {status} | {label} | {val} | {limit} |")

py_loc = m.get("source_loc_py", 0)
ts_loc = m.get("source_loc_ts", 0)
py_test = m.get("test_loc_py", 0)
ts_test = m.get("test_loc_ts", 0)
total_src = py_loc + ts_loc
total_test = py_test + ts_test

report = f"""
# Governance Metrics Report

Generated: {m.get("collected_at", "unknown")}

## Quality Gates

| Status | Metric | Current | Limit |
|--------|--------|---------|-------|
{chr(10).join(rows)}

## Codebase Size

| Metric | Python | TypeScript | Total |
|--------|--------|------------|-------|
| Source LOC | {py_loc} | {ts_loc} | {total_src} |
| Test LOC | {py_test} | {ts_test} | {total_test} |
| Mock Tax Ratio | {(py_test / py_loc):.2f}× | {(ts_test / ts_loc):.2f}× | {(total_test / total_src):.2f}× |
""".strip()

print(report)
out = Path(".memory-layer/metrics/report.md")
out.write_text(report + "\n")
print(f"\nReport written to {out}")
